### PR TITLE
Support for multiple links to a single vehicle

### PR DESCRIPTION
**Pull request by acfloria**
- Track for each link separately if it is receiving a heartbeat from a specific vehicle
- The vehicle then selects the priority link based on all active link with the following priorities
        1. USB link
        2. Non high latency links
        3. High latency links
- Any time the active property of a link for the vehicle changes, the priority link is reevaluated.
- Any time the priority link is changed from a high latency link to a regular link and vice versa a MAV_CMD_CONTROL_HIGH_LATENCY command is sent to the vehicle.
- Add a dropdown menu to manually select the priority link from all available links to the vehicle.
- Use the link name instead of the port name for SerialLinks which is also the case for all other link types.

### CODE REVIEW

**Comment by DonLakeFlyer** [comment]
I'm wondering if we can get away with not having an "Advanced Link Settings" setting. Why not just show LinkIndicator if there are multiple links and one of them is a high latency link? Removes some amount of settings clutter. Doesn't allow folks to set priority link if they don't have high latency, but not sure of the use of that and whether it would just confuse people as to what to do with it.

**Comment by DonLakeFlyer** [comment]
Can you rebase to master so I can test. There is a conflict now.

**Comment by DonLakeFlyer** [comment]
@acfloria Rebase?

**Comment by acfloria** [comment]
Yes, I will work on the rebase and your suggestions tomorrow. I was busy with other projects up to today

File: src/Vehicle/Vehicle.cc [hunk 1]
```diff
@@ -1604,7 +1597,12 @@ void Vehicle::_linkInactiveOrDeleted(LinkInterface* link)
     qCDebug(VehicleLog) << "_linkInactiveOrDeleted linkCount" << _links.count();
 
     _links.removeOne(link);
-    _updatePriorityLink();
+
+    if (_priorityLink.data() == link) {
+        _priorityLink.clear();
+    }
+
+    _updatePriorityLink(true, true);
```
[the remark this answers is not in the document]
Comment (acfloria) [reply src/Vehicle/Vehicle.cc]: True, thanks for pointing it out. Changed it.

File: src/comm/LinkInterface.h [hunk 2]
```diff
@@ -248,10 +255,24 @@ private slots:
...
+
     /// Sets the mavlink channel to use for this link
     void _setMavlinkChannel(uint8_t channel);
     
+    /**
+     * @brief timerStart
+     *
+     * Allocate the timer if it does not exist yet and start it.
+     */
+    void timerStart(int vehicle_id);
```
[the remark this answers is not in the document]
Comment (acfloria) [reply src/comm/LinkInterface.h]: Ok, changed it to `startHeartbeatTimer` and `stopHeartbeatTimer`.

File: src/Vehicle/Vehicle.cc [hunk 3]
```diff
@@ -3264,12 +3360,30 @@ void Vehicle::_vehicleParamLoaded(bool ready)
...
         _highLatencyLink = _priorityLink->highLatency();
         _mavCommandAckTimer.setInterval(_highLatencyLink ? _mavCommandAckTimeoutMSecsHighLatency : _mavCommandAckTimeoutMSecs);
         emit highLatencyLinkChanged(_highLatencyLink);
+
+        if (sendCommand) {
+            if (_highLatencyLink) {
+                sendMavCommand(defaultComponentId(),
+                               MAV_CMD_CONTROL_HIGH_LATENCY,
+                               true,
+                               1.0f); // request start transmitting over high latency telemetry
```
[the remark this answers is not in the document]
Comment (acfloria) [reply src/Vehicle/Vehicle.cc]: Changed

File: src/Vehicle/Vehicle.cc [hunk 4]
```diff
@@ -2357,43 +2437,59 @@ void Vehicle::setConnectionLostEnabled(bool connectionLostEnabled)
...
-        _maxProtoVersion = 0;
-        emit connectionLostChanged(true);
-        _say(QString(tr("%1 communication lost")).arg(_vehicleIdSpeech()));
-        if (_autoDisconnect) {
+    if (link == _priorityLink) {
+        if (active && _connectionLost) {
+            // communication to priority link regained
+            _connectionLost = false;
+            emit connectionLostChanged(false);
+            qgcApp()->showMessage((tr("%1 communication to priority link %2 regained")).arg(_vehicleIdSpeech()).arg(link->getName()));
```
[the remark this answers is not in the document]
Comment (acfloria) [reply src/Vehicle/Vehicle.cc]: Check added.

File: src/Vehicle/Vehicle.cc [hunk 5]
```diff
@@ -2057,6 +2089,54 @@ void Vehicle::setFlightMode(const QString& flightMode)
...
+
+
+    for (int i=0; i<_links.count(); i++) {
+        if (_links[i]->getName() == priorityLinkName) {
+            newPriorityLink = _links[i];
+        }
+    }
+
+    if (newPriorityLink) {
+        _priorityLinkCommanded = true;
```
[the remark this answers is not in the document]
Comment (acfloria) [reply src/Vehicle/Vehicle.cc]: I do not see a solution here as the value of `_priorityLinkCommanded` needs to be known over time and not only when `_updatePriorityLink(...)` is called.

The reason I implemented this is the following: If you switch to HL telemetry by command and you start flying away from the ground station you might loose and regain the regular telemetry thus sending a lot of high latency activate/deactivate commands over both links. By using this flag if it is commanded nothing happens if low latency telemetry is lost/regained.

File: src/comm/LinkInterface.h [hunk 6]
```diff
@@ -273,9 +294,10 @@ private slots:
     
     mutable QMutex _dataRateMutex; // Mutex for accessing the data rate member variables
 
-    bool _active;                       ///< true: link is actively receiving mavlink messages
     bool _enableRateCollection;
     bool _decodedFirstMavlinkPacket;    ///< true: link has correctly decoded it's first mavlink packet
+
+    QList<HeartbeatTimer*> _heartbeatTimerList;
```
[the remark this answers is not in the document]
Comment (acfloria) [reply src/comm/LinkInterface.h]: Sure, I changed it. Thanks for the hint.

File: src/ui/toolbar/LinkIndicator.qml [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+                    linkSelectionMenuItems.push(menuItem)
+                    linkSelectionMenu.insertItem(i, menuItem)
+                }
+            }
+        }
+        Component.onCompleted: priorityLinkSelector.updatelinkSelectionMenu()
+        Connections {
+            target:                 QGroundControl.multiVehicleManager
+            onActiveVehicleChanged: priorityLinkSelector.updatelinkSelectionMenu()
+        }
```
[the remark this answers is not in the document]
Comment (acfloria) [reply src/ui/toolbar/LinkIndicator.qml]: I tried to change it but was not able to make it work. How exactly would you do it?

**Comment by acfloria** [comment]
So rebased and addressed some comments. What is currently missing:

- Removing the advanced link settings and check instead if multiple links and at least one high latency link exist

Another question I had is should we automatically switch to the high latency link if the vehicle is not armed and low latency telemetry is lost? What is your opinion on that topic?

**Comment by DonLakeFlyer** [comment]
> Another question I had is should we automatically switch to the high latency link if the vehicle is not armed and low latency telemetry is lost?

That would seem to make sense. Although a strange case.

**Comment by acfloria** [comment]
I removed the advanced link settings as requested.
Now the link indicator is shown if at least two links are connected to the vehicle and at least one of them is a high latency one.
In addition, the menu items of inactive links are disabled which results in a light grey text and they cannot be clicked on anymore.

**Comment by acfloria** [comment]
Yes good to go from my side (tested multiple times on ground and in flight).

To test that part you don't need the hardware for the high latency link. You could for example connect a regular telemetry link and a usb connection with the high latency property set to true to a pixhawk to test it.

**Pull request merged**